### PR TITLE
Adds about:history, sets back/forward nav limit, more efficient dynamic menus

### PR DESCRIPTION
**Pull request by bsclifton**
- [x] Submitted a [ticket](https://github.com/brave/browser-laptop/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Ran `git rebase -i` to squash commits if needed.

- Brand new about:history page. It's very minimal but works (you can double click entries to launch)
- History menu now has a "Show Full History" which pulls up about:history. Fixes https://github.com/brave/browser-laptop/issues/444
- History in back/forward nav buttons is limited to 10. Fixes https://github.com/brave/browser-laptop/issues/2889
- History menu shows recently closed frames (based on WindowState)
- Broke methods out from menu.js to a helper file (menuUtil.js)
- menuUtil tests now use mockery (and work properly), stubbing electron (we can also unit test other electron dependent code like this).
- history and bookmarks menus are now dynamic and uses clear/insert calls instead of rebuilding all menu items. Fixes https://github.com/brave/browser-laptop/issues/1993
- history menu now considers modifier keys (similar to bookmarks menu)

@bridiver I could definitely use your help profiling the menu calls :smile:

For the dynamic menu, I ended up not going with removeItem...[I had made electron changes](https://github.com/bsclifton/electron/commit/13341f439ba8237316dc624805c0bf2a0ce16182), but when using the new method, electron does NOT actually remove the item from the menu object... it only hides it from the menu. Additionally, I was seeing memory leaks when I profiled this, even with electron.Menu.destroy (MenuItem does not have a destroy)

I'd also like to cc @bbondy @darkdh @diracdeltas for feedback on the changes and @alexwykoff as a heads up (I can create QA steps which outline the changes)

### CODE REVIEW

File: app/menu.js [hunk 1]
```diff
@@ -625,35 +570,75 @@ const createMenu = (CommonMenu) => {
...
+const updateMenu = (CommonMenu, appState, windowState) => {
+  const updated = menuUtil.checkForUpdate(appState, windowState)
+  if (updated.nothingUpdated) {
+    return
+  }
+
+  // Only update menu when necessary
+
+  if (updated.settings || updated.closedFrames) {
+    console.log('rebuilding history menu')
```
Comment (bridiver) [inline app/menu.js]: logs from debugging? Do we need/want these? We really need to add our own log method with more control over output granularity

Comment (bsclifton) [inline app/menu.js]: Good catch :smile: Removed!

**Comment by bbondy** [comment]
mentioned via DM, seems to be a bug with clean session when there are no sites. After navigating gto a single site it works for all future restarts.

File: js/constants/messages.js [hunk 2]
```diff
@@ -110,6 +110,9 @@ const messages = {
   RESPONSE_WINDOW_STATE: _,
   LAST_WINDOW_STATE: _,
   UNDO_CLOSED_WINDOW: _,
+  // Menu rebuilding
+  REQUEST_WINDOW_STATE_FOR_MENU: _,
+  RESPONSE_WINDOW_STATE_FOR_MENU: _,
```
Comment (bbondy) [inline js/constants/messages.js]: If you're only using closedFrames from windowState then I think you should call this:
REQUEST_WINDOW_CLOSED_FRAMES
RESPONSE_WINDOW_CLOSED_FRAMES

And then update accordingly.

Comment (bbondy) [inline js/constants/messages.js]: actually call it something like request menu data for window, and only include closed frames in it.

File: app/index.js [hunk 3]
```diff
@@ -525,7 +535,11 @@ app.on('ready', () => {
     // save app state every 5 minutes regardless of update frequency
     setInterval(initiateSessionStateSave, 1000 * 60 * 5)
     AppStore.addChangeListener(() => {
-      Menu.init(AppStore.getState().get('settings'), AppStore.getState().get('sites'))
+      if (BrowserWindow.getFocusedWindow()) {
+        BrowserWindow.getFocusedWindow().webContents.send(messages.REQUEST_WINDOW_STATE_FOR_MENU)
```
Comment (bbondy) [inline app/index.js]: I think a better place for requesting this data is in a `mainWindow.on('focus', ...` event of the `BrowserWindow` in the appStore.

File: app/extensions/brave/locales/en-US/menu.properties [hunk 4]
```diff
@@ -43,10 +43,12 @@ home=Home
 back=Back
 forward=Forward
 reopenLastClosedWindow=Reopen Last Closed Window
-showAllHistory=Show All History
+showAllHistory=Show Full History
```
Comment (bbondy) [inline app/extensions/brave/locales/en-US/menu.properties]: For now just leave out the word `Full` since app state sites data is truncated to most recent 500.

File: app/index.js [hunk 5]
```diff
@@ -295,6 +295,13 @@ app.on('ready', () => {
     saveIfAllCollected()
   })
 
+  // Window state must be fetched from main process; this is fired once it's retrieved
+  ipcMain.on(messages.RESPONSE_WINDOW_STATE_FOR_MENU, (wnd, data) => {
```
Comment (bbondy) [inline app/index.js]: pls name `data` more specific here, maybe something like `windowLevelData`

File: js/constants/messages.js (hunk 2 above)
Comment (bbondy) [inline js/constants/messages.js]: Just noticed we also have `UPDATE_APP_MENU` you should rename that to be more specific to be about bookmarks, and then maybe that is an indication that you should be more specific for perf reasons in the message names as I originally suggested in the first message in this thread.

File: js/contextMenus.js [hunk 6]
```diff
@@ -954,6 +954,8 @@ function onHamburgerMenu (location, e) {
 function onMainContextMenu (nodeProps, frame, contextMenuType) {
   if (contextMenuType === 'bookmark' || contextMenuType === 'bookmark-folder') {
     onBookmarkContextMenu(Immutable.fromJS(nodeProps), Immutable.fromJS({ location: '', title: '', partitionNumber: frame.get('partitionNumber') }))
+  } else if (contextMenuType === 'history') {
+    // TODO: add new onHistoryContextMenu() and associated methods.
```
Comment (bbondy) [inline js/contextMenus.js]: Please do this in another commit in this same PR.

File: js/contextMenus.js [hunk 7]
```diff
@@ -1072,6 +1075,18 @@ function onBackButtonHistoryMenu (activeFrame, history, rect) {
         }
       })
     }
+
+    if (stopIndex !== -1) {
+      menuTemplate.push(
+        CommonMenu.separatorMenuItem,
```
Comment (bbondy) [inline js/contextMenus.js]: I think this separator should be added conditionally on if there is at least 1 item before it. I think it's the case there could be 0.

File: js/contextMenus.js [hunk 8]
```diff
@@ -1103,6 +1119,18 @@ function onForwardButtonHistoryMenu (activeFrame, history, rect) {
         }
       })
     }
+
+    if (stopIndex !== history.entries.length) {
+      menuTemplate.push(
+        CommonMenu.separatorMenuItem,
```
Comment (bbondy) [inline js/contextMenus.js]: ditto double check on separator here.

File: js/entry.js [hunk 9]
```diff
@@ -53,6 +53,10 @@ ipc.on(messages.REQUEST_WINDOW_STATE, () => {
   ipc.send(messages.RESPONSE_WINDOW_STATE, windowStore.getState().toJS())
 })
 
+ipc.on(messages.REQUEST_WINDOW_STATE_FOR_MENU, () => {
+  ipc.send(messages.RESPONSE_WINDOW_STATE_FOR_MENU, windowStore.getState().toJS())
```
Comment (bbondy) [inline js/entry.js]: per before this is best to be more specific pls.

File: js/lib/appUrlUtil.js [hunk 10]
```diff
@@ -75,7 +75,7 @@ module.exports.isIntermediateAboutPage = (location) =>
   ['about:safebrowsing', 'about:error', 'about:certerror'].includes(getBaseUrl(location))
 
 module.exports.isNotImplementedAboutPage = (location) =>
-  ['about:config', 'about:history'].includes(getBaseUrl(location))
+  ['about:config'].includes(getBaseUrl(location))
```
Comment (bbondy) [inline js/lib/appUrlUtil.js]: yay!

File: test/unit/lib/menuUtilTest.js [hunk 11]
```diff
@@ -0,0 +1,255 @@
+/* global describe, before, after, it */
```
Comment (bbondy) [inline test/unit/lib/menuUtilTest.js]: great work on these tests.

File: js/contextMenus.js [hunk 12]
```diff
@@ -1103,6 +1119,18 @@ function onForwardButtonHistoryMenu (activeFrame, history, rect) {
...
+        {
+          label: locale.translation('showAllHistory'),
```
Comment (bbondy) [inline js/contextMenus.js]: Pls include the same item on the back and forward click and hold menus.

File: less/about/history.less [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+      padding-top: 10px;
+    }
+  }
+}
+
+.searchInput {
+    float: right;
+    padding: 5px;
+    margin-top: -35px;
+}
```
Comment (bbondy) [inline less/about/history.less]: there might be an opportunity to refactor some common styles with bookmarks page, if so please take the time to do that now before we merge this.

File: js/about/history.js [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+    return <div role='listitem'
+      className={className}
+      onContextMenu={aboutActions.contextMenu.bind(this, this.props.history.toJS(), 'history')}
+      data-context-menu-disable
+      onDoubleClick={this.navigate.bind(this)}>
+    {
+      this.props.history.get('customTitle') || this.props.history.get('title')
+      ? <span className='aboutListItem' title={this.props.history.get('location')}>
+        <span className='aboutItemDate'>{new Date(this.props.history.get('lastAccessedTime')).toLocaleDateString()}</span>
+        <span className='aboutItemTitle'>{this.props.history.get('customTitle') || this.props.history.get('title')}</span>
```
Comment (bbondy) [inline js/about/history.js]: There is a chance the title is blank, pls test and format correctly for that case.

Comment (bbondy) [inline js/about/history.js]: correctly means just with the url in that case.

File: js/about/history.js [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+          <HistoryList history={this.state.history.filter((site) => site.get('tags').isEmpty())}
+            onChangeSelectedEntry={this.onChangeSelectedEntry}
+            selectedEntry={this.state.selectedEntry} />
+        </Sticky>
+      </div>
+    </div>
+  }
+}
+
+module.exports = <AboutHistory />
```
Comment (bbondy) [inline js/about/history.js]: If you can pull out some common controls with bookmarks it would help because we plan to do things like multi select in the future. It'll be very soon work in any case so you can save the refactor for that if you prefer.

File: js/lib/menuUtil.js [hunk 16]
```diff
@@ -0,0 +1,166 @@
...
+    if (site.get('tags').includes(siteTags.BOOKMARK) && site.get('location')) {
+      payload.push({
+        // TODO include label made from favicon. It needs to be of type NativeImage
+        // which can be made using a Buffer / DataURL / local image
+        // the image will likely need to be included in the site data
+        // there was potentially concern about the size of the app state
+        // and as such there may need to be another mechanism or cache
+        //
+        // see: https://github.com/brave/browser-laptop/issues/3050
+        label: site.get('customTitle') || site.get('title'),
```
Comment (bbondy) [inline js/lib/menuUtil.js]: || site.get('location')

File: js/lib/menuUtil.js [hunk 17]
```diff
@@ -0,0 +1,166 @@
...
+          } else {
+            CommonMenu.sendToFocusedWindow(focusedWindow, [messages.SHORTCUT_ACTIVE_FRAME_LOAD_URL, site.get('location')])
+          }
+        }
+      })
+    } else if (siteUtil.isFolder(site)) {
+      const folderId = site.get('folderId')
+      const submenuItems = bookmarks.filter((bookmark) => bookmark.get('parentFolderId') === folderId)
+      payload.push({
+        label: site.get('customTitle') || site.get('title'),
```
Comment (bbondy) [inline js/lib/menuUtil.js]: || site.get('location')

**Comment by bbondy** [comment]
Looking great, I can't wait to get this in!
I made a bunch of small asks but I think we can get this in soon. Probably after the next commit.

File: js/lib/menuUtil.js (hunk 17 above)
Comment (bsclifton) [inline js/lib/menuUtil.js]: This one is a little weird, since it's a folder (and doesn't have a location). But I did implement location for the one above :smile:

**Comment by bsclifton** [comment]
OK most of the feedback implemented :smile: Next up- focusing on the concerns with the History page (breaking code into a control, implementing context menu, etc)

File: js/constants/messages.js [hunk 18]
```diff
@@ -111,8 +110,9 @@ const messages = {
   LAST_WINDOW_STATE: _,
   UNDO_CLOSED_WINDOW: _,
   // Menu rebuilding
-  REQUEST_WINDOW_STATE_FOR_MENU: _,
-  RESPONSE_WINDOW_STATE_FOR_MENU: _,
+  REQUEST_MENU_DATA_FOR_WINDOW: _,
+  RESPONSE_MENU_DATA_FOR_WINDOW: _,
+  UPDATE_MENU_BOOKMARKED_STATUS: _, /** @arg {Object} currently only has a boolean "bookmarked" */
```
Comment (bbondy) [inline js/constants/messages.js]: nit documentation bug, I think this arg is actually a boolean type and not an Object

File: js/entry.js [hunk 19]
```diff
@@ -53,6 +53,10 @@ ipc.on(messages.REQUEST_WINDOW_STATE, () => {
   ipc.send(messages.RESPONSE_WINDOW_STATE, windowStore.getState().toJS())
 })
 
+ipc.on(messages.REQUEST_MENU_DATA_FOR_WINDOW, () => {
+  ipc.send(messages.RESPONSE_MENU_DATA_FOR_WINDOW, windowStore.getState().toJS())
```
Comment (bbondy) [inline js/entry.js]: Pls send in a smaller subset here instead in a follow up

Comment (bridiver) [inline js/entry.js]: I'd really like to avoid this kind of state transfer in general and I don't think it's going to be correctly synced anyway because this updates on appState changes, not windowState changes. This will be much easier once we move webContents management to the main process and if the only purpose of this is to get close frames I would just leave it out for now. If we really need it now then I think appState should track WINDOW_CLOSE_FRAME, WINDOW_UNDO_CLOSED_FRAME. That will mean some state duplication between windowState and appState, but still better than relaying windowState imo. There is also likely a performance issue doing this on every app state change.

File: js/components/navigationBar.js [hunk 20]
```diff
@@ -51,8 +51,12 @@ class NavigationBar extends ImmutableComponent {
     if (!isBookmarked) {
       appActions.addSite(siteDetail, siteTags.BOOKMARK)
     }
+    // Show bookmarks toolbar after first bookmark is saved
     appActions.changeSetting(settings.SHOW_BOOKMARKS_TOOLBAR, !hasBookmark || showBookmarksToolbar)
+    // trigger the AddEditBookmark modal
     windowActions.setBookmarkDetail(siteDetail, siteDetail)
+    // Update checked/unchecked status in the Bookmarks menu
+    ipc.send(messages.UPDATE_MENU_BOOKMARKED_STATUS, this.bookmarked)
```
Comment (bridiver) [inline js/components/navigationBar.js]: similar to the issue with DATA_FOR_WINDOW, we're using ipc messages in a lot of places where we could (and should I think) just be using actions. The windowActions are already sent to the appStore via the dispatcher, we're just not handling them there. Ultimately this seems a little backwards to me anyway because bookmarks are part of appState, not windowState and the bookmarked state of the current url should come from there. I realize there are legacy reasons for all of this, but it would really be nice to clean some of it up.

File: js/lib/menuUtil.js [hunk 21]
```diff
@@ -0,0 +1,166 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public * License, v. 2.0. If a copy of the MPL was not distributed with this file,
```
Comment (bridiver) [inline js/lib/menuUtil.js]: new files should use the new directory structure - see https://github.com/brave/browser-laptop/blob/master/docs/directoryStructure.md

**Comment by bbondy** [comment]
++

File: js/lib/menuUtil.js (hunk 21 above)
Comment (bsclifton) [inline js/lib/menuUtil.js]: sweet, didn't know about this- thanks :smile:

Comment (bridiver) [inline js/lib/menuUtil.js]: yea, I guess @bbondy and I talked about it and then didn't tell anyone :)

**Comment by echosa** [comment]
@bsclifton Does this PR also make reopened tabs have the same history they had when closed? https://github.com/brave/browser-laptop/issues/2702

**Comment by bsclifton** [comment]
@echosa great catch- no, it does not.  It seems that #444 was used to not only track an about:history page but also restoring the history for a reopened tab.  I went ahead and re-opened https://github.com/brave/browser-laptop/issues/2702 and can take a look at that while I'm in the history code :smile:

**Comment by echosa** [comment]
@bsclifton Great! Thanks! This is one of my (ever-dwindling) checklist items of "must be implemented before Brave becomes my default browser". :-D Looking forward to it! :+1:

**Comment by bridiver** [comment]
it does not, but I was going to add in the necessary methods in electron to make that functionality possible so we can add it in the next few releases
